src/syntax.rs

```diff
@@ -0,0 +1,302 @@
+            ]
+        })
+        .boxed()
+    }
+}
+
+impl<F: LurkField> fmt::Display for Syntax<F> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::Num(_, x) => write!(f, "{}", x),
+            Self::UInt(_, x) => write!(f, "{}u64", x),
+            Self::Symbol(_, sym) => write!(f, "{}", sym),
+            Self::LurkSym(_, sym) => write!(f, "{}", sym),
+            Self::String(_, x) => write!(f, "\"{}\"", x.escape_default()),
+            Self::Char(_, x) => {
+                if *x == '(' || *x == ')' {
+                    write!(f, "'\\{}'", x)
+                } else {
+                    write!(f, "'{}'", x.escape_default())
+                }
+            }
+            Self::Quote(_, x) => write!(f, "'{}", x),
+            Self::List(_, xs) => {
+                let mut iter = xs.iter().peekable();
+                write!(f, "(")?;
+                while let Some(x) = iter.next() {
+                    match iter.peek() {
+                        Some(_) => write!(f, "{} ", x)?,
+                        None => write!(f, "{}", x)?,
+                    }
+                }
+                write!(f, ")")
+            }
+            Self::Improper(_, xs, end) => {
+                let mut iter = xs.iter().peekable();
+                write!(f, "(")?;
+                while let Some(x) = iter.next() {
+                    match iter.peek() {
+                        Some(_) => write!(f, "{} ", x)?,
+                        None => write!(f, "{} . {}", x, *end)?,
+                    }
+                }
+                write!(f, ")")
+            }
+        }
+    }
+}
+
+impl<F: LurkField> Store<F> {
+    pub fn intern_syntax(&mut self, syn: Syntax<F>) -> Ptr<F> {
+        match syn {
+            Syntax::Num(_, x) => self.intern_num(x),
+            Syntax::UInt(_, x) => self.intern_uint(x),
+            Syntax::Char(_, x) => self.intern_char(x),
+            Syntax::Symbol(_, x) => self.intern_symbol(x),
+            Syntax::LurkSym(_, x) => self.intern_symbol(Symbol::lurk_sym(&format!("{x}"))),
+            Syntax::String(_, x) => self.intern_string(x),
+            Syntax::Quote(pos, x) => {
+                let xs = vec![Syntax::Symbol(pos, Symbol::lurk_sym("quote")), *x];
+                self.intern_syntax(Syntax::List(pos, xs))
+            }
+            Syntax::List(_, xs) => {
+                let mut cdr = self.nil();
+                for x in xs.into_iter().rev() {
+                    let car = self.intern_syntax(x);
+                    cdr = self.intern_cons(car, cdr);
+                }
+                cdr
+            }
+            Syntax::Improper(_, xs, end) => {
+                let mut cdr = self.intern_syntax(*end);
+                for x in xs.into_iter().rev() {
+                    let car = self.intern_syntax(x);
+                    cdr = self.intern_cons(car, cdr);
+                }
+                cdr
+            }
+        }
+    }
+
```

Comment:
What do you think of
```
    /// Tries to fetch a, possibly improper, list in a syntactically canonical form. The empty list `()` is translated into
    /// symbol `nil`. A list of 2 elements of form `(quote x)`, i.e. `(quote . (x . ()))`, becomes `'x`. Otherwise,
    /// the list is either a proper list, ending in `Expression::Nil` or an improper list
```
for an explanation? I'm trying to follow CL's way of printing stuff (apart from capitalizing). Scheme's approach to quote is exactly the same as well. If you type `'(quote x)` in the repl you get `'x` (or `'X`), and `'(quote x y)` you get `(quote x y)` (or `(QUOTE X Y)`)

Reply:
Fair enough! The comment helps, thanks!

Reply:
I actually couldn't follow the comment. I think if the goal is to explain proper vs improper lists, then the explanation of `quote`, etc. makes it hard to follow. Also, I if you describe `()` as being 'the empty list', then it is not *translated* to `nil`. Rather, both `()` and `nil` are examples of syntax which represent the symbol `nil`, which is also defined to be the empty list.

Reply:
```
    /// Tries to fetch a syntactic list from an expression pointer, by looping over cons cells and
    /// collecting their contents. If the ptr does not point to a cons or nil (i.e. not a list) we
    /// return None. If after traversing zero or more cons cells we hit a `nil`, we return a proper
    /// list (`Syntax::List`), otherwise an improper list (`Syntax::Improper`). If the proper list
    /// is a quotation `(quote x)`, then we return the syntactic quotation `Syntax::Quote`
```

which clarifies that what we are doing here is returning `Syntax` from an `Expression` (represented by `Ptr` into a `Store`)